File: 0542-01-matrix/0542-01-matrix.java
class Solution {
    public int[][] updateMatrix(int[][] mat) {
        Queue<int[]> queue = new LinkedList<>();
        int[][] result = new int[mat.length][mat[0].length];

        for (int i = 0; i < mat.length; i++) {
            for (int j = 0; j < mat[i].length; j++) {
                if (mat[i][j] == 0) {
                    queue.offer(new int[]{i, j});
                    result[i][j] = 0;
                } else {
                    result[i][j] = Integer.MAX_VALUE;
                }
            }
        }

        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int row = cur[0], col = cur[1];

            for (int[] dir : dirs) {
                int dr = row + dir[0];
                int dc = col + dir[1];

                if (dr >= 0 &&
                    dr < mat.length &&
                    dc >= 0 &&
                    dc < mat[0].length &&
                    result[dr][dc] > result[row][col] + 1) {
                    result[dr][dc] = result[row][col] + 1;
                    queue.offer(new int[]{dr, dc});
                }
            }
        }

        return result;
    }
}

File: 0542-01-matrix/0542-01-matrix.kt
class Solution {
    fun updateMatrix(mat: Array<IntArray>): Array<IntArray> {
        val queue = ArrayDeque<Pair<Int, Int>>()
        val result = Array(mat.size) { IntArray(mat[0].size) }

        for (i in mat.indices) {
            for (j in mat[i].indices) {
                if (mat[i][j] == 0) {
                    queue.add(i to j)
                    result[i][j] = 0
                } else {
                    result[i][j] = Int.MAX_VALUE
                }
            }
        }

        val dirs = arrayOf(-1 to 0, 1 to 0, 0 to -1, 0 to 1)

        while (queue.isNotEmpty()) {
            val (row, col) = queue.removeFirst()

            for ((dr, dc) in dirs) {
                val nr = row + dr
                val nc = col + dc

                if (nr in mat.indices &&
                    nc in mat[0].indices &&
                    result[nr][nc] > result[row][col] + 1) {
                    result[nr][nc] = result[row][col] + 1
                    queue.add(nr to nc)
                }
            }
        }

        return result
    }
}